README
Welcome to the BeeCrypt crypto library!

Copyright (c) 1997, 1998, 1999, 2000, 2001, 2004, 2005 X-Way Rights BV
Copyright (c) 2002, 2003, 2005, 2006 Bob Deblier

Author: Bob Deblier <bob.deblier@telenet.be>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For the specifics of this license, see file 'COPYING.LIB', included in
this distribution.



Welcome to version 4.2.0 of BeeCrypt:

The C++ API has received a major overhaul and has been expanded to deal
properly with multi-threading on the major platforms.

About BeeCrypt:

BeeCrypt started its life when the need for a portable and fast cryptography
library arose at Virtual Unlimited in 1997. I'm still trying to make it
faster, easier to use and more portable, in addition to providing better
documentation.

Note that depending on where you are, the use of cryptography may be
limited or forbidden by law. Before using this library, make sure you
are legally entitled to do so.

<plug>
Most of the algorithms are implemented from reliable sources such as:

"Handbook of Applied Cryptography"
 Alfred J. Menezes, Paul C. van Oorschot, Scott A. Vanstone
 CRC Press

"Applied Cryptography", second edition
 Bruce Schneier
 Wiley


For crypto enthusiasts these books are invaluable background material.

IEEE P1363 "Standard Specifications for Public Key Cryptography" is a
very interesting draft standard, which we will try to comply with.
</plug>

The structures in the library are geared towards exchange with Java
and its security and cryptography classes. This library can also be
accessed from Java by installing BeeCrypt for Java, a JCE 1.2 crypto
provider and the counterpart of this library.


Included in the library are:
    - entropy sources for initializing pseudo-random generators
    - pseudo-random generators: FIPS-186, Mersenne Twister
    - block ciphers: AES, Blowfish
    - hash functions: MD5, SHA-1, SHA-256, SHA-384, SHA-512
    - keyed hash functions: HMAC-MD5, HMAC-SHA-1, HMAC-SHA-256, HMAC-SHA-384, HMAC-SHA-512
    - multi-precision integer library, with assembler-optimized routines
      for several processors
    - probabilistic primality testing, with optimized small prime trial
      division
    - discrete logarithm parameter generation over a prime field
    - Diffie-Hellman key agreement
    - DSA signature scheme
    - ElGamal signature scheme (two variants)
    - RSA keypair generation with chinese remainder theorem variables
    - RSA public & private key operations
    - DHAES encryption scheme

Planned for the near future are:
    - compliance with and compliance statements for IEEE P1363
    - more blockciphers (Twofish, ... )
    - more blickcipher mods (OFB, ... )
    - more hash functions (MD2, MD4, Tiger)
    - Elliptic Curves (ECDSA, ... )
    - RSA signatures as specified by RFC-2440.

The library has been tested on the following platforms:
    - Cygwin
    - Darwin/MacOS X
    - Linux glibc 2.x alpha
    - Linux glibc 2.x arm
    - Linux glibc 2.x ia64
    - Linux glibc 2.x m68k
    - Linux glibc 2.x ppc
    - Linux glibc 2.x ppc64
    - Linux glibc 2.x s390
    - Linux glibc 2.x s390x
    - Linux glibc 2.x sparc
    - Linux glibc 2.x x86
    - Linux glibc 2.x x86_64/amd64
    - Solaris 2.[6789] sparc (with Forte or gnu compilers)
    - Solaris 2.[78] x86 (with Forte or GNU compilers)
    - Tru64 Unix alpha
    - Win32 (Windows 95, 98, NT 4.0, 2000, XP)

The library is currently in the process of being ported to:
    - MinGW
    - AIX (shared libraries don't seem to work in 64-bit mode)

For more information, refer to the HTML documentation, which can be generated
with Doxygen, in the docs directory.

If you want to report bugs, make suggestions, contribute fixes or
enhancements, please see the beecrypt-specific website:

http://sourceforge.net/projects/beecrypt

or contact me at mailto:bob.deblier@telenet.be

Sincerely,

Bob Deblier


NEWS
4.2.2:
	- Added support for pkgconfig
	- Fixed python path lookup in configure.ac
	- Fixed compilation problem on MinGW
	- Fixed C++ interface / implementation const differences in parameters

4.2.1:
	- Cleaned up old autotools cruft (thanks to Anìbal Monsalve Salazar).
	- Different automake approach to compiling separate components.
	- Fixed C++ header installation problem.

4.2.0:
	- Added OpenMP support (initially for RSA CRT).
	- Added SHA-224 and HMAC-SHA-224.
	- Added RIPEMD-128/160/256/320 (thanks to Jeff Johnson).
	- Added MD4 (thanks to Jeff Johnson).
	- Applied RedHat/Debian/Gentoo patches for 4.1.2.
	- Improved gcc compilation flags for sparc.
	- Started work on MinGW compatibility.
	- Major overhaul of the C++ API; be warned that the code is still
	  tricky in certain places.
	- Added synchronization methods to Object class; there's also a
	  macro allowing the use of synchronized (...) { }.
	- Added Runnable interface and Thread class; the latter provides
	  basic support conform to the Java API.
	- Added collections.

4.1.2:
	- Fixed Cygwin DLL missing symbols problem.
	- Fixed GNU-stack assembler section on some platforms (Debian-ARM).
	- Fixed problem cause by include of <asm/byteorder.h>.
	- Fixed SHA-384 and SHA-512 code Visual C++ compatibility.
	- Improved detection of IBM ICU library version; has to be >= 2.8.

4.1.1:
	- Fixed shared library version info.

4.1.0:
	- Added SHA-384 and SHA-512 algorithms.
	- Added HMAC-SHA-384 and HMAC-SHA-512 algorithms.
	- Added generic SSE2 optimization for the above algorithms.
	- Added more digest algorithms for PKCS#1 EMSA.
	- Optimized swap32 and swap64 routines on Linux.
	- Fixed missing definition in mpopt.h for s390x.
	- Fixed nostackexec configuration bug.
	- Fixed problem in Date::toString.
	- Fixed deadlock problem which occured in certain cases where security
	  or crypto SPI constructor called getInstance for another security or
	  crypto SPI.
	- Fixed a bug in the generic CBC encryption code; when called with
	  nblocks == 1, the feedback was set incorrectly.
	- Fixed a bug in mpbsubmod; sometimes it takes multiple additions of
	  the modulus to get a positive number.
	- Fixed PowerPC 64-bit configuration problem on Linux.

4.0.0:
	- Added a C++ API interface, modeled after Java's security & crypto API.
	- Added the new GNU noexecstack feature.
	- Added more x86_64 and s390x assembler routines.
	- Modified i2osp, so that it only requires as many octets as there are
	  significant bytes in the multi-precision integers.
	- Fixed a bug in the creation of rsa keypairs; code was not correctly
	  migrated to new calling sequence. The code now implements the method
	  described in IEEE P.1363.
	- Fixed another bug in mpextgcd_w which sometimes returned incorrect
	  results.
	- Fixed a bug in mprshiftlsz, which didn't work correctly when size = 1.
	- Fixed a configuration problem on Tru64 Unix.

3.1.0:
	- Added wiping of private key components of keypairs before freeing.
	- Fixed bug in mpextgcd_w which sometimes returned incorrect result.
	- Fixed error in PowerPC 64-bit assembler symbol definitions.

3.0.0:
	- Cleaned up installed header files.
	- Modified the API so that all keys can be passed as arrays of bytes.
	- Modified the API so that all key sizes are given in bits.
	- Modified the multi-precision integer library to work better on 64-bit
	  machines.
	- Modified the assembly source generation mechanism, employing the m4
	  macro processor.
	- Added multi-precision integer vectorized assembler routines for
	  Itanium.
	- Added multi-precision integer assembler routines for PowerPC 64-bit.
	- Added multi-precision integer assembler routines for Alpha.
	- Added multi-precision integer assembler routines for Opteron.
	- Added multi-precision integer assembler routines for IBM zSeries 64-bit.
	- Added multi-precision integer assembler routines for M68K.
	- Added Jeff Johnson's python bindings.
	- Added new unit tests.
	- Added new benchmarking programs.

2.3.0pre:
	- Modified the header files so that the library now uses self-contained autoconf-generated configuration files; a program employing BeeCrypt can now use the symbols already tested and defined instead of having to regenerate them (thus also eliminating the risk of inconsistencies).
	- Added the AES algorithm, with assembler routines for i586 and powerpc.
	- Added the DSA signature algorithm.
	- Added PowerPC assembler routines for blowfish.
	- Added Pentium4 SSE2 assembler multiplication routines.
	- Fixed the RSA CRT algorithm.
	- Fixed the gas/i386 mp32even and mp32odd routines.
	- Fixed a bug in modular inverse computation; thanks to Jeff Johnson of RedHat for pointing this out.
	- Fixed a bug in testing the result of a gcd operation in the mp32prndconone routine.
	- Fixed an ugly bug in base64 decoding.
	- Fixed compatibility with the latest automake & autoconf versions.
	- Replaces CPU optimization mechanism in configure script.

2.1.0:
	- Added support for automake, autoheader and libtool, which should make compiling the library even easier.
	- Changed DHAES API to conform to IEEE P.1363 submission and to allow for uneven key splitting.
	- Improved PKCS#5 padding routines.
	- Added a hash reset to the hashFunctionContextInit function. This was pointed out by Marko Kreen.
	- Fixed problem with configuring on i486-pc-linux-gnu. This was pointed out Steve O'Neill.
	- Fixed problem in the C version of mp32sub where carry would sometimes be missed. This was pointed out by Jon Sturgeon.
	- Revised entropy gathering system to do timeouts & asynchronous I/O where possible, to avoid hangs in case there's no noise on the audio device (i.e. digital silence), or when no data is available on devices such as /dev/random.
	- Changed mp32opt i386 assembler routines for slight performance improvement.
	- Changed mp32opt powerpc assembler routines for slight performance improvement.
	- Changed mp32opt sparcv9 assembler routines for slight performance improvement.

dsa.h
/*
 * Copyright (c) 2001, 2002 Beeyond Software Holding BV
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

/*!\file dsa.h
 * \brief Digital Signature Algorithm, as specified by NIST FIPS 186.
 *
 * FIPS 186 specifies the DSA algorithm as having a large prime \f$p\f$,
 * a cofactor \f$q\f$ and a generator \f$g\f$ of a subgroup of
 * \f$\mathds{Z}^{*}_p\f$ with order \f$q\f$. The private and public key
 * values are \f$x\f$ and \f$y\f$ respectively.
 *
 * \author Bob Deblier <bob.deblier@telenet.be>
 * \ingroup DL_dsa_m
 */

#ifndef _DSA_H
#define _DSA_H

#include "beecrypt/dlkp.h"

typedef dldp_p dsaparam;
typedef dlpk_p dsapub;
typedef dlkp_p dsakp;

#ifdef __cplusplus
extern "C" {
#endif

/*!\fn int dsasign(const mpbarrett* p, const mpbarrett* q, const mpnumber* g, randomGeneratorContext* rgc, const mpnumber* hm, const mpnumber* x, mpnumber* r, mpnumber* s)
 * \brief This function performs a raw DSA signature.
 *
 * Signing equations:
 *
 * \li \f$r=(g^{k}\ \textrm{mod}\ p)\ \textrm{mod}\ q\f$
 * \li \f$s=k^{-1}(h(m)+xr)\ \textrm{mod}\ q\f$
 *
 * \param p The prime.
 * \param q The cofactor.
 * \param g The generator.
 * \param rgc The pseudo-random generator context.
 * \param hm The hash to be signed.
 * \param x The private key value.
 * \param r The signature's \e r value.
 * \param s The signature's \e s value.
 * \retval 0 on success.
 * \retval -1 on failure.
 */
BEECRYPTAPI
int dsasign(const mpbarrett* p, const mpbarrett* q, const mpnumber* g, randomGeneratorContext*, const mpnumber* hm, const mpnumber* x, mpnumber* r, mpnumber* s);

/*!\fn int dsavrfy(const mpbarrett* p, const mpbarrett* q, const mpnumber* g, const mpnumber* hm, const mpnumber* y, const mpnumber* r, const mpnumber* s)
 * \brief This function performs a raw DSA verification.
 *
 * Verifying equations:
 * \li Check \f$0<r<q\f$ and \f$0<s<q\f$
 * \li \f$w=s^{-1}\ \textrm{mod}\ q\f$
 * \li \f$u_1=w \cdot h(m)\ \textrm{mod}\ q\f$
 * \li \f$u_2=rw\ \textrm{mod}\ q\f$
 * \li \f$v=(g^{u_1}y^{u_2}\ \textrm{mod}\ p)\ \textrm{mod}\ q\f$
 * \li Check \f$v=r\f$
 *
 * \param p The prime.
 * \param q The cofactor.
 * \param g The generator.
 * \param hm The digest to be verified.
 * \param y The public key value.
 * \param r The signature's \e r value.
 * \param s The signature's \e s value.
 * \retval 1 on success.
 * \retval 0 on failure.
 */
BEECRYPTAPI
int dsavrfy(const mpbarrett* p, const mpbarrett* q, const mpnumber* g, const mpnumber* hm, const mpnumber* y, const mpnumber* r, const mpnumber* s);

/*!\fn int dsaparamMake(dsaparam* dp, randomGeneratorContext* rgc, size_t psize)
 * \brief This function generates a set of DSA parameters.
 *
 * This function calls dldp_pgoqMake with appropriate parameters, i.e.
 * qsize  = 160 bits and cofactor = 1.
 *
 * \param dp The parameters to be generated.
 * \param rgc The random generator context.
 * \param psize The size of prime parameter p; psize must be >= 512 and <= 1024, and be a multiple of 64.
 * \retval 0 on success.
 * \retval -1 on failure.
 */
BEECRYPTAPI
int dsaparamMake(dsaparam*, randomGeneratorContext*, size_t);

#ifdef __cplusplus
}
#endif

#endif
